### PR TITLE
feat: implement syntax highlighting for preview text

### PR DESCRIPTION
**Pull request by cartercanedy**
image

Tree-sitter is a permissively-licensed, open-source parser framework, well known for its integrations in many popular editors (i use nvim, btw). Luckily, they provide Rust bindings that are reasonably straightforward to use. This PR implements syntax highlighting for the source preview, making it much easier to read for people wanting to understand the code behind commands that they are running.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Documentation update
- [x] Refactoring
- [ ] Hotfix
- [ ] Security patch
- [x] UI/UX improvement

## Description
- Implement tree-sitter-based syntax highlighting

## Testing
Manual testing by perusing the previews of various commands

## Impact
This should make it much easier to understand the source while being presented in the preview

## Checklist
- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no errors/warnings/merge conflicts.

### CODE REVIEW

File: tui/src/floating_text.rs [hunk 1]
```diff
@@ -1,56 +1,159 @@
...
+}
+
+macro_rules! style {
+    ($r:literal, $g:literal, $b:literal) => ({
+        use anstyle::{Color, Style, RgbColor};
+        Style::new().fg_color(Some(Color::Rgb(RgbColor($r, $g, $b))))
+    })
+}
+
+const SYNTAX_HIGHLIGHT_STYLES: [(&'static str, anstyle::Style); 8] = [
```
Comment (adamperkowski) [inline tui/src/floating_text.rs:44]:
```suggestion
const SYNTAX_HIGHLIGHT_STYLES: [(&str, anstyle::Style); 8] = [
```

File: tui/src/floating_text.rs [hunk 2]
```diff
@@ -1,56 +1,159 @@
...
+    ("string",      style!(206, 145, 120)), // brown
+    ("property",    style!(156, 220, 254)), // light blue
+    ("comment",     style!(92,  131, 75)),  // green
+    ("embedded",    style!(206, 145, 120)), // blue (string expansions)
+    ("constant",    style!(79,  193, 255)), // dark blue
+    ("keyword",     style!(197, 134, 192)), // magenta
+    ("number",      style!(181, 206, 168))  // light green
+];
+
+fn get_highlighted_string<'a>(s: &'a str) -> Option<String> {
```
Comment (adamperkowski) [inline tui/src/floating_text.rs:55]:
```suggestion
fn get_highlighted_string(s: &str) -> Option<String> {
```

**Review by adamperkowski — CHANGES_REQUESTED** [review]
Awesome.
Just remember to run `cargo fmt --all`.

**Comment by cartercanedy** [comment]
> Just remember to run `cargo fmt --all`.

:))

**Review by adamperkowski — APPROVED** [review]

**Comment by adamperkowski** [comment]
> and remove Cargo.lock

Don't remove it if you're adding deps.
If it doesn't conflict, then it's fine.

**Comment by cartercanedy** [comment]
> > and remove Cargo.lock
> 
> Don't remove it if you're adding deps.
> If it doesn't conflict, then it's fine.

Did the Cargo.lock ci check get removed?

**Comment by adamperkowski** [comment]
> Did the Cargo.lock ci check get removed?

It did because `Cargo.lock` exists for some reason.
Just don't push a deleated and re-cleanbuilded `Cargo.lock` if not neccessary.

**Comment by cartercanedy** [comment]
>Just don't push a deleated and re-cleanbuilded `Cargo.lock` if not neccessary

makes sense

**Comment by Real-MullaC** [comment]
Please take a look at the check.

**Comment by cartercanedy** [comment]
> Please take a look at the check.

already on it

**Comment by lj3954** [comment]
This causes a reversion by removing functionality added in #202.

**Comment by cartercanedy** [comment]
Yeah, that'll have to get sorted out. I'll have to take a deeper look into breaking up the lines on overflow, since it gets a little complicated with the unrendered ansi sequences in the raw strings

**Comment by cartercanedy** [comment]
Just perusing the docs a bit, this shouldn't be too hard. I'd probably still use a flat map, just over the constructed lines rather than the raw text to preserve the styling and accurately determine the width of the actually rendered text

**Comment by cartercanedy** [comment]
@lj3954 why are we looking at wrapping rather than side scrolling?

**Comment by ChrisTitusTech** [comment]
> @lj3954 why are we looking at wrapping rather than side scrolling?

Oh I like side scrolling a lot better.

**Comment by cartercanedy** [comment]
> Oh I like side scrolling a lot better. 

Me 2, but there's also the issue of altered semantics of a mid line break without also inserting a talking backslash, and even then I'm not sure that backslashes are a perfect solution

**Comment by cartercanedy** [comment]
@ChrisTitusTech @lj3954 just implemented side scrolling to fix regression

**Review by adamperkowski — APPROVED** [review]
`v_scroll` and `h_scroll` could be renamed to `scroll_v` and `scroll_h` but it's not that big.

**Comment by lj3954** [comment]
get_lines should not take in an owned String, the split method is implemented for str instead. The lines() method also is more idiomatic than split("\n").

**Comment by cartercanedy** [comment]
> get_lines should not take in an owned String, the split method is implemented for str instead. The lines() method also is more idiomatic than split("\n").

good catch, that was pretty sloppy

**Review by ChrisTitusTech — APPROVED** [review]

**Pull request merged**